### PR TITLE
[Feature] Support TAM

### PR DESCRIPTION
**Pull request by congee524**
Support TAM

### CODE REVIEW

**Comment by congee524** [comment]
This model is almost ready. Please review it when you are free.

**Comment by innerlee** [comment]
Cool! @dreamerlin if you have time, please take a first pass on the code :)

File: configs/recognition/tanet/tanet_r50_1x1x8_50e_sthv1_rgb.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    client_cfg='/mnt/lustre/share/memcached_client/client.conf',
+    sys_path='/mnt/lustre/share/pymc/py3')
+
+sthv1_flip_label_map = {2: 4, 4: 2, 30: 41, 41: 30, 52: 66, 66: 52}
+img_norm_cfg = dict(
+    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_bgr=False)
+
+train_pipeline = [
+    dict(type='SampleFrames', clip_len=1, frame_interval=1, num_clips=8),
+    dict(type='RawFrameDecode', io_backend='memcached', **mc_cfg),
```
Comment (dreamerlin) [inline configs/recognition/tanet/tanet_r50_1x1x8_50e_sthv1_rgb.py:27]: memcached should not be set as default

File: configs/recognition/tanet/tanet_r50_1x1x8_50e_sthv1_rgb.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    dict(type='ToTensor', keys=['imgs', 'label'])
+]
+val_pipeline = [
+    dict(
+        type='SampleFrames',
+        clip_len=1,
+        frame_interval=1,
+        num_clips=8,
+        test_mode=True),
+    dict(type='RawFrameDecode', io_backend='memcached', **mc_cfg),
```
Comment (dreamerlin) [inline configs/recognition/tanet/tanet_r50_1x1x8_50e_sthv1_rgb.py:50]: Same

File: configs/recognition/tanet/tanet_r50_1x1x8_50e_sthv1_rgb.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+]
+data = dict(
+    videos_per_gpu=8,
+    workers_per_gpu=4,
+    test_dataloader=dict(videos_per_gpu=4),
+    train=dict(
+        type=dataset_type,
+        ann_file=ann_file_train,
+        data_prefix=data_root,
+        filename_tmpl='{:05}.jpg',
```
Comment (dreamerlin) [inline configs/recognition/tanet/tanet_r50_1x1x8_50e_sthv1_rgb.py:82]: remove it

File: configs/recognition/tanet/tanet_r50_dense_1x1x8_100e_kinetics400_rgb.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+data_root = 'data/kinetics400/rawframes_train'
+data_root_val = 'data/kinetics400/rawframes_val'
+ann_file_train = 'data/kinetics400/kinetics400_train_list_rawframes.txt'
+ann_file_val = 'data/kinetics400/kinetics400_val_list_rawframes.txt'
+ann_file_test = 'data/kinetics400/kinetics400_val_list_rawframes.txt'
+
+mc_cfg = dict(
+    server_list_cfg='/mnt/lustre/share/memcached_client/server_list.conf',
+    client_cfg='/mnt/lustre/share/memcached_client/client.conf',
+    sys_path='/mnt/lustre/share/pymc/py3')
```
Comment (dreamerlin) [inline configs/recognition/tanet/tanet_r50_dense_1x1x8_100e_kinetics400_rgb.py:16]: redundant

File: configs/recognition/tanet/tanet_r50_1x1x8_50e_sthv2_rgb.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+data_root = 'data/sthv2/rawframes'
+data_root_val = 'data/sthv2/rawframes'
+ann_file_train = 'data/sthv2/sthv2_train_list_rawframes.txt'
+ann_file_val = 'data/sthv2/sthv2_val_list_rawframes.txt'
+ann_file_test = 'data/sthv2/sthv2_val_list_rawframes.txt'
+
+mc_cfg = dict(
+    server_list_cfg='/mnt/lustre/share/memcached_client/server_list.conf',
+    client_cfg='/mnt/lustre/share/memcached_client/client.conf',
+    sys_path='/mnt/lustre/share/pymc/py3')
```
Comment (dreamerlin) [inline configs/recognition/tanet/tanet_r50_1x1x8_50e_sthv2_rgb.py:19]: redundant

File: configs/recognition/tanet/tanet_r50_1x1x8_50e_sthv1_rgb.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+data_root = 'data/sthv1/rawframes'
+data_root_val = 'data/sthv1/rawframes'
+ann_file_train = 'data/sthv1/sthv1_train_list_rawframes.txt'
+ann_file_val = 'data/sthv1/sthv1_val_list_rawframes.txt'
+ann_file_test = 'data/sthv1/sthv1_val_list_rawframes.txt'
+
+mc_cfg = dict(
+    server_list_cfg='/mnt/lustre/share/memcached_client/server_list.conf',
+    client_cfg='/mnt/lustre/share/memcached_client/client.conf',
+    sys_path='/mnt/lustre/share/pymc/py3')
```
Comment (dreamerlin) [inline configs/recognition/tanet/tanet_r50_1x1x8_50e_sthv1_rgb.py:19]: redunant

File: mmaction/models/backbones/tanet.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+import torch.nn as nn
+from torch.utils import checkpoint as cp
+
+from ..common import TAM
+from ..registry import BACKBONES
+from .resnet import Bottleneck, ResNet
+
+
+class TABlock(nn.Module):
+    """Temporal Adaptive Block (TA-Block) for TANet.
```
Comment (dreamerlin) [inline mmaction/models/backbones/tanet.py:12]: Paper link should be added like https://github.com/open-mmlab/mmaction2/blob/master/mmaction/models/backbones/resnet3d_slowfast.py#L366-L367

File: mmaction/models/backbones/tanet.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+    "The temporal adaptive module (TAM) is embedded into ResNet-Block
+    after the first Conv2D, which turns the vanilla ResNet-Block
+    into TA-Block."
```
Comment (dreamerlin) [inline mmaction/models/backbones/tanet.py:16]: quotation is redundant

File: mmaction/models/common/tam.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
+import torch.nn as nn
+import torch.nn.functional as F
+from mmcv.cnn import constant_init, kaiming_init, normal_init
+
+
+class TAM(nn.Module):
+    """Temporal Adaptive Module(TAM) for TANet.
+
+    https://arxiv.org/pdf/2005.06803.pdf.
```
Comment (dreamerlin) [inline mmaction/models/common/tam.py:9]: The list is not in rst form, please refer to https://github.com/open-mmlab/mmaction2/blob/master/mmaction/models/backbones/resnet3d_slowfast.py#L366-L367

File: mmaction/models/common/tam.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+                 in_channels,
+                 num_segments,
+                 alpha=2,
+                 adaptive_kernel_size=3,
+                 beta=4,
+                 conv1d_kernel_size=3,
+                 adaptive_convolution_stride=1,
+                 adaptive_convolution_padding=1,
+                 init_std=0.001):
+        super(TAM, self).__init__()
```
Comment (dreamerlin) [inline mmaction/models/common/tam.py:43]: ~~TAM, self~~

**Comment by dreamerlin** [comment]
unittest in `test_recognizer2d.py` should also be added

File: configs/recognition/tanet/tanet_r50_1x1x8_50e_sthv1_rgb.py [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        pipeline=test_pipeline))
+evaluation = dict(
+    interval=1, metrics=['top_k_accuracy', 'mean_class_accuracy'])
+
+# optimizer
+optimizer = dict(
+    type='SGD',
+    constructor='TSMOptimizerConstructor',
+    paramwise_cfg=dict(fc_lr5=True),
+    lr=0.02,  # this lr is used for 8 gpus
```
[the remark this answers is not in the document]
Comment (congee524) [reply configs/recognition/tanet/tanet_r50_1x1x8_50e_sthv1_rgb.py:97]: TANet performs better with setting `lr=0.02`. I'm testing it on sthv1.

**Comment by congee524** [comment]
kindly ping :)

File: configs/recognition/tanet/tanet_r50_1x1x8_50e_sthv1_rgb.py [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+    dict(
+        type='SampleFrames',
+        clip_len=1,
+        frame_interval=1,
+        num_clips=8,
+        twice_sample=True,
+        test_mode=True),
+    dict(type='RawFrameDecode'),
+    dict(type='Resize', scale=(-1, 256)),
+    # dict(type='CenterCrop', crop_size=224),
```
Comment (dreamerlin) [inline configs/recognition/tanet/tanet_r50_1x1x8_50e_sthv1_rgb.py:63]: Remove this

File: configs/recognition/tanet/tanet_r50_1x1x8_50e_sthv1_rgb.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (dreamerlin) [reply configs/recognition/tanet/tanet_r50_1x1x8_50e_sthv1_rgb.py:97]: OK

File: configs/recognition/tanet/tanet_r50_dense_1x1x8_100e_kinetics400_rgb.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+test_pipeline = [
+    dict(
+        type='DenseSampleFrames',
+        clip_len=1,
+        frame_interval=1,
+        num_clips=8,
+        test_mode=True),
+    dict(type='RawFrameDecode'),
+    dict(type='Resize', scale=(-1, 256)),
+    # dict(type='CenterCrop', crop_size=224),
```
Comment (dreamerlin) [inline configs/recognition/tanet/tanet_r50_dense_1x1x8_100e_kinetics400_rgb.py:58]: Remove this

File: mmaction/models/common/tam.py [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+            complexity in the local branch. Default: 4.
+        conv1d_kernel_size (int): Size of the convolution kernel of Conv1d in
+            the local branch. Default: 3.
+        adaptive_convolution_stride (int): The first dimension of strides in
+            the adaptive convolution of ```Temporal Adaptive Aggregation```.
+            Default: 1.
+        adaptive_convolution_padding (int): The first dimension of paddings in
+            the adaptive convolution of ```Temporal Adaptive Aggregation```.
+            Default: 1.
+        init_std (float): Std value for initiation of nn.Linear. Default:
```
Comment (dreamerlin) [inline mmaction/models/common/tam.py:30]: ``nn.Linear``

Comment (congee524) [reply mmaction/models/common/tam.py:30]: mmmm, I don't understand.

**Comment by congee524** [comment]
@innerlee Maybe @dreamerlin has done it. Could you take a moment to review the code?

**Comment by innerlee** [comment]
Okay, soon

File: configs/_base_/models/tanet_r50.py [hunk 15]
```diff
@@ -0,0 +1,19 @@
+# model settings
```
Comment (innerlee) [inline configs/_base_/models/tanet_r50.py:1]: double check the filename, `ta...net`?

Comment (innerlee) [reply configs/_base_/models/tanet_r50.py:1]: okay fine

**Review by innerlee — APPROVED** [review]

**Comment by innerlee** [comment]
Please modify changelog and repo's README later

**Pull request merged**